gdcdatamodel/validators/link_validator_parser.py

```diff
@@ -0,0 +1,185 @@
+        update_mask_should_fix = child.required
+        self.code = child.code if self.code is None else self.code | child.code
+        if update_mask_should_fix:
+            self.existing_mask = child.code if self.existing_mask is None else self.existing_mask | child.code
+            self.existing_links.append(child.name)
+
+        current_sg = self
+        if not current_sg.required:
+            update_mask_should_fix = False
+        if current_sg.required:
+            current_sg.required_links.append(child.name)
+
+        while current_sg is not None:
+            parent = current_sg.parent
+            if parent is not None:
+                if update_mask_should_fix:
+                    update_mask_should_fix = self.update_existing_mask(current_sg, parent, child, update_mask_should_fix)
+                parent.code = current_sg.code if parent.code is None else parent.code | current_sg.code
+                if parent.required:
+                    parent.required_links.append(child.name)
+            current_sg = parent
+        self.update_exclusive_mask(child)
+
+    def add_child_link(self, child):
+        self.update_parent(child)
+
+
+class LinkWithExclusiveMask(object):
+    def __init__(self, item):
+        self.name = item.name
+        self.exclusive_mask = item.exclusive_mask
+        self.exclusive_links = item.exclusive_links
+        self.multiplicity = item.schema.get('multiplicity')
+        self.backref = item.schema.get('backref')
+
+
+class ExistingMasks(object):
+    def __init__(self, code, existing_mask, existing_links):
+        self.code = code
+        self.existing_mask = existing_mask
+        self.existing_links = existing_links
+
+
+class RequiredMask(object):
+    def __init__(self):
+        self.list_required_links = []
+        self.group_required = []
+        self.required_mask = 0
+
+
+def create_subgroup_validators(node_label):
+    link_items = []
+    leaves = []
+    for link in gdcdictionary.schema[node_label]['links']:
+        if 'name' in link:
+            l_item = SubgroupParsingNode(None, link, 1 << len(leaves))
+            leaves.append(l_item)
+        else:
+            l_item = SubgroupParsingNode(None, link)
+        link_items.append(l_item)
+
+    return build_tree_of_link_items(link_items, leaves)
+
+
+def build_tree_of_link_items(link_items, leaves):
+    final_required = []
+    current_pos = 0
+    while current_pos < len(link_items):
+        l_item = link_items[current_pos]
+        children_required_count = 0
+        if 'subgroup' in l_item.schema:
+            for link in l_item.schema['subgroup']:
+                if 'subgroup' in link:
+                    new_l_item = SubgroupParsingNode(l_item, link)
+                else:
+                    new_l_item = SubgroupParsingNode(l_item, link, 1 << len(leaves))
+                    l_item.add_child_link(new_l_item)
+                    leaves.append(new_l_item)
+
+                link_items.append(new_l_item)
```

Comment:
I think it would be clearer to separate this into 2 loops, first one to build up the complete `link_items` list and the second one to append `SubgroupParsingNode`s to the leaves list? I know it's pretty much the same, I just feel like it might be clearer to build up the list to iterate first rather than mutate it while looping with the `while`